ddtrace/tracer/abandoned_spans.go

```diff
@@ -0,0 +1,154 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package tracer
+
+import (
+	"container/list"
+	"fmt"
+	"strings"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+var tickerInterval = time.Minute
+var logSize = 9000
+
+// reportAbandonedSpans periodically finds and reports potentially abandoned
+// spans that are older than the given interval. These spans are stored in a
+// bucketed linked list, sorted by their `Start` time.
+func (t *tracer) reportAbandonedSpans(interval time.Duration) {
+	tick := time.NewTicker(tickerInterval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+			logAbandonedSpans(t.abandonedSpans, true, interval)
+		case s := <-t.cIn:
+			bNode := t.abandonedSpans.Front()
+			if bNode == nil {
+				b := list.New()
+				b.PushBack(s)
+				t.abandonedSpans.PushBack(b)
+				break
+			}
+			for bNode != nil {
+				bucket, bOk := bNode.Value.(*list.List)
+				if !bOk {
+					bNode = bNode.Next()
+					continue
+				}
+				sNode := bucket.Front()
+				if sNode == nil {
+					bNode = bNode.Next()
+					continue
+				}
+				sp, sOk := sNode.Value.(*span)
+				if sOk && sp != nil && s.Start-sp.Start <= interval.Nanoseconds() {
+					bucket.PushBack(s)
+					break
+				}
+				bNode = bNode.Next()
+			}
+			if bNode != nil {
+				break
+			}
+			b := list.New()
+			b.PushBack(s)
+			t.abandonedSpans.PushBack(b)
+		case s := <-t.cOut:
+			for node := t.abandonedSpans.Front(); node != nil; node = node.Next() {
+				bucket, ok := node.Value.(*list.List)
+				if !ok || bucket.Front() == nil {
+					continue
+				}
+				spNode := bucket.Front()
+				sp, ok := spNode.Value.(*span)
+				if !ok {
+					continue
+				}
+				if s.Start-sp.Start <= interval.Nanoseconds() {
+					bucket.Remove(spNode)
+					if bucket.Front() == nil {
+						t.abandonedSpans.Remove(node)
+					}
+					break
+				}
+			}
+		case <-t.stop:
+			logAbandonedSpans(t.abandonedSpans, false, interval)
+			return
+		}
+	}
+}
+
+// logAbandonedSpans returns a string containing potentially abandoned spans. If `filter` is true,
+// it will only return spans that are older than the provided time `interval`. If false,
+// it will return all unfinished spans.
+func logAbandonedSpans(l *list.List, filter bool, interval time.Duration) {
+	var sb strings.Builder
+	nowTime := now()
+	spanCount := 0
+	truncated := false
+
+	for bucketNode := l.Front(); bucketNode != nil; bucketNode = bucketNode.Next() {
+		bucket, ok := bucketNode.Value.(*list.List)
+		if !ok || bucket == nil {
+			continue
+		}
```

Comment:
Since I see that you do this kind of type checking a couple of times for bucket nodes and span nodes, maybe consider making it a helper fn in this file. You can try it out, and if it doesn't help readability, don't worry about it.